pkg/reduce/reduce_test.go

```diff
@@ -544,6 +544,65 @@ func TestReduceDataForward_SumWithDifferentKeys(t *testing.T) {
 
 }
 
+func TestDataForward_WithContextClose(t *testing.T) {
+	var (
+		ctx, cancel    = context.WithTimeout(context.Background(), 5*time.Second)
+		fromBufferSize = int64(100000)
+		toBufferSize   = int64(10)
+		messages       = []int{100, 99}
+		startTime      = 0 // time in millis
+		fromBufferName = "source-reduce-buffer"
+		toBufferName   = "reduce-to-buffer"
+		err            error
+	)
+
+	defer cancel()
+
+	// create from buffers
+	fromBuffer := simplebuffer.NewInMemoryBuffer(fromBufferName, fromBufferSize)
+
+	// create to buffers
+	buffer := simplebuffer.NewInMemoryBuffer(toBufferName, toBufferSize)
+	toBuffer := map[string]isb.BufferWriter{
+		toBufferName: buffer,
+	}
+
+	// create a store provider
+	storeProvider := memory.NewMemoryStores(memory.WithStoreSize(1000))
+
+	// create pbq manager
+	var pbqManager *pbq.Manager
+	pbqManager, err = pbq.NewManager(ctx, storeProvider,
+		pbq.WithReadTimeout(1*time.Second), pbq.WithChannelBufferSize(10))
+	assert.NoError(t, err)
+
+	// create in memory watermark publisher and fetcher
+	f, p := fetcherAndPublisher(ctx, toBuffer, fromBuffer, t.Name())
+
+	// create a fixed window of 5 minutes
+	window := fixed.NewFixed(5 * time.Minute)
+
+	var reduceDataForward *DataForward
+	reduceDataForward, err = NewDataForward(ctx, SumReduceTest{}, fromBuffer, toBuffer, pbqManager, CounterReduceTest{}, f, p,
+		window, WithReadBatchSize(1))
+	assert.NoError(t, err)
+
+	// start the forwarder
+	go reduceDataForward.Start(ctx)
+	// publish messages with 1 millisecond delay
+	go publishMessages(ctx, startTime, messages, 600, 10, p[fromBuffer.GetName()], fromBuffer, 1*time.Millisecond)
+	// sleep for 100 milliseconds, so that few messages are processed and partitions are created
+	time.Sleep(100 * time.Millisecond)
+	cancel()
+	// wait for a second for the messages to get persisted
+	time.Sleep(time.Second)
```

Comment:
do we really need this sleep? i think we should put a `for` loop to check whether there are some messages

Reply:
yes removed sleep